### PR TITLE
Validator to submit attestations when seeing new head

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

+ We want to submit attestation as long as validator see new head or 1/3 slot, whichever comes first

**Description**

+ this does not resolve the missing head metric of validator monitor completely, just an improvement
+ need to work on having subnet peers in order to resolve that, see #3348

### CODE REVIEW

File: packages/lodestar/src/chain/validation/attestation.ts [hunk 1]
```diff
@@ -172,12 +172,14 @@ export async function validateGossipAttestation(
  * Note: We do not queue future attestations for later processing
  */
 export function verifyPropagationSlotRange(chain: IBeaconChain, attestationSlot: Slot): void {
+  // don't want to have the check for dev network with small SECONDS_PER_SLOT
+  const maximumGossipClockDisparity = chain.config.SECONDS_PER_SLOT <= 3 ? 0 : MAXIMUM_GOSSIP_CLOCK_DISPARITY_SEC;
```
Comment (dapplion) [inline packages/lodestar/src/chain/validation/attestation.ts]: Why is this necessary? Does this break the sim tests?

File: packages/validator/src/services/attestation.ts [hunk 2]
```diff
@@ -42,8 +44,10 @@ export class AttestationService {
     // Fetch info first so a potential delay is absorved by the sleep() below
     const dutiesByCommitteeIndex = groupAttDutiesByCommitteeIndex(this.dutiesService.getDutiesAtSlot(slot));
 
-    // Lighthouse recommends to always wait to 1/3 of the slot, even if the block comes early
-    await sleep(this.clock.msToSlotFraction(slot, 1 / 3), signal);
+    // A validator should create and broadcast the attestation to the associated attestation subnet when either
+    // (a) the validator has received a valid block from the expected block proposer for the assigned slot or
+    // (b) one-third of the slot has transpired (SECONDS_PER_SLOT / 3 seconds after the start of slot) -- whichever comes first.
+    await Promise.race([sleep(this.clock.msToSlotFraction(slot, 1 / 3), signal), this.waitForBlockSlot(slot)]);
```
Comment (dapplion) [inline packages/validator/src/services/attestation.ts:50]: What if the block has already been seen before calling waitForBlockSlot()? You could check this edge case with the head tracker

File: packages/lodestar/src/chain/validation/attestation.ts (hunk 1 above)
Comment (twoeths) [reply packages/lodestar/src/chain/validation/attestation.ts]: yeah otherwise it breaks e2e and sim tests since validators submit attestations too early

File: packages/validator/src/services/attestation.ts (hunk 2 above)
Comment (twoeths) [reply packages/validator/src/services/attestation.ts:50]: `waitForBlockSlot` does use the head tracker, this tracks the block from the slot starts so should be safe enough

File: packages/lodestar/src/chain/validation/attestation.ts (hunk 1 above)
Comment (dapplion) [reply packages/lodestar/src/chain/validation/attestation.ts]: I understand but it feels too arbitrary to have this here this way. With too early, do you mean the beacon node is busy and it's internal clock is too slow?

Comment (twoeths) [reply packages/lodestar/src/chain/validation/attestation.ts]: no, the clock is right but due to the MAXIMUM_GOSSIP_CLOCK_DISPARITY_SEC (0.5s), it can only accept attestations of current slot after `clockSlot + 0.5s` while this PR make validator submit attestations within `clockSlot` and `clockSlot + 0.5s` which is too early

Comment (twoeths) [reply packages/lodestar/src/chain/validation/attestation.ts]: @dapplion thanks for pointing that out, it's a bug. I fixed it by implementing `slotWithFutureTolerance` and `slotWithPastTolerance` to make it clear

Comment (twoeths) [reply packages/lodestar/src/chain/validation/attestation.ts]: merged via #3409

**Review by dapplion — APPROVED** [review]
